token/src/token.rs

```diff
@@ -0,0 +1,173 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use zkvm::{
+    Commitment, Data, Entry, Input, Instruction, Predicate, Program, Prover, Signature, Tx,
+    TxHeader, TxID, TxLog, VMError, Value, VerificationKey, Verifier,
+};
+
+/// Represents a ZkVM Token with unique flavor and embedded
+/// metadata protected by a user-supplied Predicate.
+pub struct Token {
+    issuance_predicate: Predicate,
+    metadata: Vec<u8>,
+}
+
+impl Token {
+    /// Constructs a new Token.
+    pub fn new(pred: Predicate, metadata: Vec<u8>) -> Self {
+        Token {
+            issuance_predicate: pred,
+            metadata: metadata,
+        }
+    }
+
+    /// Returns the Token's flavor.
+    pub fn flavor(&self) -> Scalar {
+        Value::issue_flavor(&self.issuance_predicate, Data::Opaque(self.metadata.clone()))
+    }
+
+    /// Returns program that issues specified quantity of Token.
+    pub fn issue<'a>(program: &'a mut Program, token: &Token, qty: u64) -> &'a mut Program {
+        program
+            .push(Commitment::blinded(qty)) // stack: qty
+            .var() // stack: qty-var
+            .push(Commitment::unblinded(token.flavor())) // stack: qty-var, flv
+            .var() // stack: qty-var, flv-var
+            .push(Data::Opaque(token.metadata.clone())) // stack: qty-var, flv-var, data
+            .push(token.issuance_predicate.clone()) // stack: qty-var, flv-var, data, flv-pred
+            .issue() // stack: issue-contract
+            .sign_tx() // stack: issued-value
+    }
+
+    /// Returns program that issues specified quantity of Token,
+    /// outputting it to the destination Predicate.
+    pub fn issue_to<'a>(
+        program: &'a mut Program,
+        token: &Token,
+        qty: u64,
+        dest: &Predicate,
+    ) -> &'a mut Program {
+        Token::issue(program, token, qty)
+            .push(dest.clone())
+            .output(1)
+    }
+
+    /// Returns program that retires a Token.
+    /// TBD: better symmetry here to accept a Token object
+    pub fn retire<'a>(
+        program: &'a mut Program,
+        qty: CompressedRistretto,
+        flv: CompressedRistretto,
+        pred: &Predicate,
```

Comment:
move in `pred` instead of cloning a borrowed one. https://rust-lang-nursery.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control